### ISSUE TITLE
KC5.2.1 Integration of KC4 and KC5 to enable the execution of a workflow of Toil/Chronos/PIVOT

### ISSUE THREAD

**Issue by clarisca** [heliumdatacommons/MVP-repo#16]
- [ ] Demonstrate execution of TOIL workflow on  one cloud in PIVOT (via Chronos deployed as part of PIVOT framework)

**Comment by dcvan24** [comment]
We have successfully run the sample TOIL workflow on top of a Chronos/Mesos cluster deployed in Azure. Thanks to @theferrit32, TOIL is able to submit jobs to any custom Chronos/Mesos cluster by properly defining environment variables `TOIL_CHRONOS_ENDPOINT` and `TOIL_CHRONOS_PROTO`.

**Comment by theferrit32** [comment]
Recent changes:
- Toil is now dockerized on worker nodes, instead of requiring a pre-setup step of putting toil on them.
- iRods password auth is automated through environment variable injection to the docker run command.

Container and toil changes documented here:

https://github.com/theferrit32/toil/issues/1

**Comment by dcvan24** [comment]
- Chronos has been deployed aside Marathon on the Azure cluster in PIVOT
- An appliance for running the sample TOIL workflow has been created and tested with the latest dockerized TOIL (theferrit32/toil#1).

Issue:
- The workflow will be repeatedly launched since Marathon is designed for running long-running jobs/services. A deeper integration with Chronos may be needed for handling one-off tasks.

**Comment by clarisca** [comment]
Absolutely. The SciDAS API should be expanded to support Chronos as a Mesos framework.

**Comment by dcvan24** [comment]
I played around with the Chronos API to get a feeling about how much effort is needed to extend SciDAS API to support Chronos.  The Chronos API is very similar as the Marathon's except the following:

- Chronos does not have a `portMapping` field for specifying port mappings between containers and the host. However, since Chronos allows users to pass in Docker options in the `parameters` field, port mappings can be done by passing in the `-p` option. 

- It can be problematic to create the container network using Weave Net **if each job needs to be assigned an IP address**. The jobs tend to fail if there is an IP address conflict. In Marathon, the conflict will be eventually resolved by restarting the failed service/app (although it can take many attempts), which is always done by Marathon. However, in Chronos, I do not find a way to retry a failed job infinitely, so it is possible that the conflict will not be resolved when the maximum number of retries is reached. 

The second issue only matters in the case where there are parallel jobs (on different Mesos workers) need to communicate with each other via the container network, which can be done alternatively using a shared FS. If it is assumed that each workflow job runs independently, there will not be any issue with the container network.

**Comment by theferrit32** [comment]
Issue on the toil fork that addresses future work on supporting multiple chronos endpoints that can span cloud systems, and also assign individual workflow tasks to be run on specific chronos endpoints. 

https://github.com/theferrit32/toil/issues/3

**Comment by dcvan24** [comment]
- Extended PIVOT API to support one-off jobs: bd7a9861a8ece65fc1c00074436b4c3bccf4865c
- Ran TOIL workflow on PIVOT in a single Cloud (Azure): 4af1bb07ef180eb7a6c86a354b0dea96be90002a
- @theferrit32 and I fixed an open bug in Chronos: https://github.com/mesos/chronos/issues/568
